File: src/uk/co/hgreen/hexclockwallpaper/HexclockWallpaperService.java
package uk.co.hgreen.hexclockwallpaper;

import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.Paint.Align;
import android.graphics.Typeface;
import android.os.Handler;
import android.service.wallpaper.WallpaperService;
import android.text.format.Time;
import android.view.SurfaceHolder;

public class HexclockWallpaperService extends WallpaperService {

	@Override
	public Engine onCreateEngine() {
		return new MyWallpaperEngine();
	}

	private class MyWallpaperEngine extends Engine {
		private final Handler handler = new Handler();
		private final Runnable drawRunner = new Runnable() {
			@Override
			public void run() {
				draw();
			}

		};
		private Paint paint = new Paint();
		private boolean visible = true;
		private Time time;
		private Typeface font;

		public MyWallpaperEngine() {
			font =  Typeface.createFromAsset(getApplicationContext().getAssets(), "fonts/Lato-Hairline.ttf");
			paint.setAntiAlias(true);
			paint.setColor(Color.WHITE);
			paint.setTextSize(120.0f);
			paint.setTextAlign(Align.CENTER);
			paint.setTypeface(font);
			handler.post(drawRunner);
			time = new Time(Time.getCurrentTimezone());
		}

		@Override
		public void onVisibilityChanged(boolean visible) {
			this.visible = visible;
			if (visible) {
				handler.post(drawRunner);
			} else {
				handler.removeCallbacks(drawRunner);
			}
		}

		@Override
		public void onSurfaceDestroyed(SurfaceHolder holder) {
			super.onSurfaceDestroyed(holder);
			this.visible = false;
			handler.removeCallbacks(drawRunner);
		}

		@Override
		public void onSurfaceChanged(SurfaceHolder holder, int format,
				int width, int height) {
			super.onSurfaceChanged(holder, format, width, height);
		}

		private void draw() {
			SurfaceHolder holder = getSurfaceHolder();
			Canvas canvas = null;
			
			try {
				canvas = holder.lockCanvas();
				if (canvas != null) {
					drawTime(canvas);
				}
			} finally {
				if (canvas != null)
					holder.unlockCanvasAndPost(canvas);
			}
			handler.removeCallbacks(drawRunner);
			if (visible) {
				handler.postDelayed(drawRunner, 50);
			}
		}
		
		private void drawTime(Canvas canvas){
			int xPos = (canvas.getWidth() / 2);
			int yPos = (int) ((canvas.getHeight() / 2) - ((paint.descent() + paint.ascent()) / 2)) ; 
			time.setToNow();
			String current = time.format("%k%M%S");
			int hex = Integer.parseInt(current, 16);
			int r = (hex >> 16) & 0xFF;
			int g = (hex >> 8) & 0xFF;
			int b = (hex >> 0) & 0xFF;
			canvas.drawRGB(r, g, b);
			canvas.drawText("#"+current, xPos, yPos-60, paint);
		}
	}

}

File: app/src/main/kotlin/uk/co/hgreen/hexclockwallpaper/HexclockWallpaperService.kt
package uk.co.hgreen.hexclockwallpaper

import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.graphics.Paint.Align
import android.graphics.Typeface
import android.os.Handler
import android.service.wallpaper.WallpaperService
import android.view.SurfaceHolder
import java.text.SimpleDateFormat
import java.util.Calendar

class HexclockWallpaperService : WallpaperService() {

    override fun onCreateEngine(): WallpaperService.Engine {
        return WallpaperEngine()
    }

    private inner class WallpaperEngine : WallpaperService.Engine() {
        private val handler = Handler()
        private val drawRunner = Runnable { draw() }
        private val paint = Paint()
        private var width: Int = 0
        private var height: Int = 0
        private var visible = true
        private val dateFormat: SimpleDateFormat
        private val font: Typeface

        init {
            font = Typeface.createFromAsset(applicationContext.assets, "fonts/Lato-Hairline.ttf")
            paint.isAntiAlias = true
            paint.color = Color.WHITE
            paint.textSize = 120.0f
            paint.textAlign = Align.CENTER
            paint.typeface = font
            handler.post(drawRunner)
            dateFormat = SimpleDateFormat("HHmmss")
        }

        override fun onVisibilityChanged(visible: Boolean) {
            this.visible = visible
            if (visible) {
                handler.post(drawRunner)
            } else {
                handler.removeCallbacks(drawRunner)
            }
        }

        override fun onSurfaceDestroyed(holder: SurfaceHolder) {
            super.onSurfaceDestroyed(holder)
            this.visible = false
            handler.removeCallbacks(drawRunner)
        }

        override fun onSurfaceChanged(holder: SurfaceHolder, format: Int, width: Int, height: Int) {
            this.width = width
            this.height = height
            super.onSurfaceChanged(holder, format, width, height)
        }

        private fun draw() {
            val holder = surfaceHolder
            var canvas: Canvas? = null

            try {
                canvas = holder.lockCanvas()
                if (canvas != null) {
                    drawTime(canvas)
                }
            } finally {
                if (canvas != null) {
                    holder.unlockCanvasAndPost(canvas)
                }
            }

            handler.removeCallbacks(drawRunner)

            if (visible) {
                handler.postDelayed(drawRunner, 50)
            }
        }

        private fun drawTime(canvas: Canvas) {
            val xPos = width / 2
            val yPos = (height / 2 - (paint.descent() + paint.ascent()) / 2).toInt()
            val current = dateFormat.format(Calendar.getInstance().time)
            val hex = Integer.parseInt(current, 16)
            val r = hex shr 16 and 0xFF
            val g = hex shr 8 and 0xFF
            val b = hex shr 0 and 0xFF
            canvas.drawRGB(r, g, b)
            canvas.drawText("#" + current, xPos.toFloat(), (yPos - 60).toFloat(), paint)
        }
    }

}